### ISSUE TITLE
Overlapping unions creates issues with inference

### ISSUE THREAD

**Issue by yilinwei** [python/mypy#7579]
- version: 7.3.0

When creating a union which has overlapping types, the inference fails. Either this should throw an error, or it should be inferred correctly.

```python
from typing import TypeVar, Union, Iterable, List

T = TypeVar("T")

def foo(test: Union[T, List[T]]):
    pass

foo([1]) # Fails with incompatible type List[<nothing>]

```

- Example of misuse [here](https://github.com/marshmallow-code/marshmallow/pull/1416)

**Comment by ilevkivskyi** [comment]
Duplicate of https://github.com/python/mypy/issues/6417
